src/words/ast/INodeRetrieveProperty.java

```diff
@@ -9,8 +9,54 @@ public INodeRetrieveProperty(Object... children) {
 	}
 
 	@Override
+	/**
+	 * Returns the value of the second child (identifier).
+	 * If the first child(reference list) evaluates to NOTHING, the identifier must be an object.
+	 * Otherwise, the reference list evaluates to an object, and identifier refers to one of the object's properties.
+	 * Returns NOTHING if the property isn't found.
+	 * Throws a WordsReferenceException if the reference_list or identifier doesn't refer to an object when it's expected to.
+	 */
 	public ASTValue eval(WordsEnvironment environment) throws WordsRuntimeException {
-		// TODO
-		throw new AssertionError("Not yet implemented");
+		ASTValue refList = children.get(0).eval(environment);
+		if (refList.type == ASTValue.ValueType.NOTHING) {
+			ASTValue id = children.get(1).eval(environment);
+			
+			try {
+				WordsObject obj = environment.getObject(id.stringValue);
+				return new ASTValue(obj);
+			} catch (WordsObjectNotFoundException e) {
+				throw new WordsReferenceException("Value expression is not an object");
+			}
+		}
+		
+		WordsObject obj = refList.objValue;
+		if (obj == null) {
+			throw new WordsReferenceException("Value expression is not an object");
+		}
+		
+		ASTValue id = children.get(1).eval(environment);
+		String propName = id.stringValue;
+		
+		WordsProperty wordsProp = obj.getProperty(propName);
```

Comment:
Since this code is basically the same as in INodeAssign, I think it would be better to make it a  method of the class.  For example, the ASTValue class could have a method called "toWordsProperty()" or something like that which does this conversion.